src/seq/index.rs

```diff
@@ -387,37 +387,49 @@ where
 
     impl<N> Eq for Element<N> {}
 
-    let mut candidates = Vec::with_capacity(length.as_usize());
+    let mut candidates = Vec::with_capacity(amount.as_usize());
     let mut index = N::zero();
-    while index < length {
+    while index < length && candidates.len() < amount.as_usize() {
         let weight = weight(index.as_usize()).into();
         if weight > 0.0 {
-            let key = rng.random::<f64>().powf(1.0 / weight);
+            let key = rng.random::<f64>().ln() / weight;
+            // let key = rng.random::<f64>().powf(1.0 / weight);
             candidates.push(Element { index, key });
         } else if !(weight >= 0.0) {
             return Err(WeightError::InvalidWeight);
         }
 
         index += N::one();
     }
+    candidates.sort_unstable();
 
-    let avail = candidates.len();
-    if avail < amount.as_usize() {
+    if candidates.len() < amount.as_usize() {
         return Err(WeightError::InsufficientNonZero);
     }
 
-    // Partially sort the array to find the `amount` elements with the greatest
-    // keys. Do this by using `select_nth_unstable` to put the elements with
-    // the *smallest* keys at the beginning of the list in `O(n)` time, which
-    // provides equivalent information about the elements with the *greatest* keys.
-    let (_, mid, greater) = candidates.select_nth_unstable(avail - amount.as_usize());
+    let mut x = rng.random::<f64>().ln() / candidates[0].key;
+    while index < length {
+        let weight = weight(index.as_usize()).into();
+        if weight > 0.0 {
+            x -= weight;
+            if x <= 0.0 {
+                let t = core::f64::consts::E.powf(candidates[0].key * weight);
+                let key = rng.random_range(t..1.0).ln() / weight;
+                candidates[0] = Element { index, key };
+                candidates.sort_unstable();
```

Comment:
I guess it is very likely that a tree data structure would perform much faster if amount is big enough. Not sure where the threshold is. Depending on `sort_unstable` is could even perform particularly worse on an almost sorted slice.

Reply:
Good point, though I won't address it now.

Reply:
https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html
should be always faster as it also stores its elements in a `Vec` should be an easy change.